proxy/src/auth/backend/jwt.rs

```diff
@@ -0,0 +1,549 @@
+use std::{future::Future, sync::Arc, time::Duration};
+
+use anyhow::{bail, ensure, Context};
+use arc_swap::ArcSwapOption;
+use bytes::Bytes;
+use dashmap::DashMap;
+use jose_jwk::crypto::KeyInfo;
+use serde::de::DeserializeOwned;
+use signature::Verifier;
+use tokio::time::Instant;
+
+use crate::intern::EndpointIdInt;
+
+#[derive(Default)]
+pub struct JWKCache {
+    client: reqwest::Client,
+
+    map: DashMap<EndpointIdInt, Arc<JWKCacheEntryLock>>,
+}
+
+pub struct JWKCacheEntryLock {
+    cached: ArcSwapOption<JWKCacheEntry>,
+    lookup: tokio::sync::Semaphore,
+}
+
+impl Default for JWKCacheEntryLock {
+    fn default() -> Self {
+        JWKCacheEntryLock {
+            cached: ArcSwapOption::empty(),
+            lookup: tokio::sync::Semaphore::new(1),
+        }
+    }
+}
+
+pub struct JWKCacheEntry {
+    /// Should refetch at least every hour to verify when old keys have been removed.
+    /// Should refetch when new key IDs are seen only every 5 minutes or so
+    last_retrieved: Instant,
+
+    // /// jwks urls
+    // urls: Vec<url::Url>,
+    /// cplane will return multiple JWKs urls that we need to scrape.
+    key_sets: ahash::HashMap<url::Url, jose_jwk::JwkSet>,
+}
+
+/// How to get the JWT auth rules
+pub trait FetchAuthRules: Clone + Send + Sync + 'static {
+    fn fetch_auth_rules(&self) -> impl Future<Output = anyhow::Result<AuthRules>> + Send;
+}
+
+#[derive(Clone)]
+struct FetchAuthFromCplane {
+    #[allow(dead_code)]
+    endpoint: EndpointIdInt,
+}
+
+impl FetchAuthRules for FetchAuthFromCplane {
+    async fn fetch_auth_rules(&self) -> anyhow::Result<AuthRules> {
+        Err(anyhow::anyhow!("not yet implemented"))
+    }
+}
+
+pub struct AuthRules {
+    jwks_urls: Vec<url::Url>,
+}
+
+impl JWKCacheEntryLock {
+    async fn acquire_permit(&self) -> DetachedPermit {
+        match self.lookup.acquire().await {
+            Ok(permit) => {
+                permit.forget();
+                DetachedPermit
+            }
+            Err(_) => panic!("semaphore should not be closed"),
+        }
+    }
+
+    fn try_acquire_permit(&self) -> Option<DetachedPermit> {
+        match self.lookup.try_acquire() {
+            Ok(permit) => {
+                permit.forget();
+                Some(DetachedPermit)
+            }
+            Err(tokio::sync::TryAcquireError::NoPermits) => None,
+            Err(tokio::sync::TryAcquireError::Closed) => panic!("semaphore should not be closed"),
+        }
+    }
+
+    async fn renew_jwks<F: FetchAuthRules>(
```

Comment:
I think this method should be split into what belongs to the Lock type and the business logic that belongs to the Entry type. Or call the former Entry and the latter EntryInner if that makes more sense. Up to you.